plugins/src/gcc/imp.rs

```diff
@@ -0,0 +1,1270 @@
+            self.variance = (1. - MOVING_AVERAGE_SMOOTHING_FACTOR)
+                * (self.variance + MOVING_AVERAGE_SMOOTHING_FACTOR * avg_diff * avg_diff);
+            self.standard_dev = self.variance.sqrt();
+        }
+    }
+
+    fn estimate_is_close(&self, value: Bitrate) -> bool {
+        if self.average == 0. {
+            false
+        } else {
+            ((self.average - STANDARD_DEVIATION_CLOSE_NUM * self.standard_dev)
+                ..(self.average + STANDARD_DEVIATION_CLOSE_NUM * self.standard_dev))
+                .contains(&(value as f64))
+        }
+    }
+}
+
+struct State {
+    /// Note: The target bitrate applied is the min of
+    /// target_bitrate_on_delay and target_bitrate_on_loss
+    estimated_bitrate: Bitrate,
+
+    /// Bitrate target based on delay factor for all video streams.
+    /// Hasn't been tested with multiple video streams, but
+    /// current design is simply to divide bitrate equally.
+    target_bitrate_on_delay: Bitrate,
+
+    /// Used in additive mode to track last control time, influences
+    /// calculation of added value according to gcc section 5.5
+    last_increase_on_delay: Option<time::Instant>,
+    last_decrease_on_delay: time::Instant,
+
+    /// Bitrate target based on loss for all video streams.
+    target_bitrate_on_loss: Bitrate,
+
+    last_increase_on_loss: time::Instant,
+    last_decrease_on_loss: time::Instant,
+
+    /// Exponential moving average, updated when bitrate is
+    /// decreased
+    ema: ExponentialMovingAverage,
+
+    last_control_op: BandwidthEstimationOp,
+
+    min_bitrate: Bitrate,
+    max_bitrate: Bitrate,
+
+    detector: Detector,
+
+    periodic_push: Option<gst::PeriodicClockId>,
+
+    // Implemented like a leaky bucket
+    buffers: VecDeque<gst::Buffer>,
+
+    flow_return: Result<gst::FlowSuccess, gst::FlowError>,
+    last_push: time::Instant,
+}
+
+impl State {
+    fn create_buffer_list(&mut self, bwe: &super::BandwidthEstimator) -> gst::BufferList {
+        let now = time::Instant::now();
+        let elapsed = Duration::from_std(now - self.last_push).unwrap();
+        let mut budget = (elapsed.num_nanoseconds().unwrap())
+            .mul_div_round(
+                self.estimated_bitrate as i64,
+                gst::ClockTime::SECOND.nseconds() as i64,
+            )
+            .unwrap();
+        let total_budget = budget;
+        let mut remaining = self.buffers.iter().map(|b| b.size() as f64).sum::<f64>() * 8.;
+        let total_size = remaining;
+
+        let mut list = gst::BufferList::new();
+        let mutlist = list.get_mut().unwrap();
+
+        // Leak the bucket so it can hold at most 30ms of data
+        let maximum_remaining_bits = 30. * self.estimated_bitrate as f64 / 1000.;
+        while (budget > 0 || remaining > maximum_remaining_bits) && !self.buffers.is_empty() {
+            let buf = self.buffers.pop_back().unwrap();
+            let n_bits = buf.size() * 8;
+
+            mutlist.add(buf);
```

Comment:
Does this rely on basesink/nicesink not handling waiting for each buffer in a buffer list?

Reply:
The synchronization should happen upstream indeed, and the sink should not sync anything.

In `webrtcsink` this is how we do it, but indeed, if the stream is not synced on the clock before webrtcbin it is gonna work because of the suboptimal synchronization  behavior of  buffer lists in BaseSink.

Reply:
Maybe this should come together with a change to split bufferlists somewhere in webrtcbin or webrtcsink

Reply:
I am not sure to understand your point.

Reply:
If it's a problem that buffer lists are synchronized as one item in sinks instead of pacing each buffer individually then you probably want to make sure that buffer lists are split into individual buffers before the synchronization happens (in the `clocksync` elements here, the sinks instead `webrtcbin` don't sync!).

Reply:
Well ideally upstream synchronizes before the buffer are put into a list and then we do our pacing without any more synchronization, this is what we do in `webrtcsink`, adding support to sync on individual buffers in `clocksync` could be a good idea too but unrelated to me.

Would you want to add a `clocksync` into `webrtcbin` automatically when using a congestion controller? I do not think it is a good idea. I guess the documentation of `rtpgccbwe` should mention those details though.

Reply:
`webrtcbin` already uses `clocksync` internally to synchronize the separate streams before bundling, but that would still work on full buffer lists at this point.

Reply:
Ah right, well, not sure it is a problem, especially if the stream has already been synced upstream. Adding proper support to bufferlist sync in clocksync/basesink is anyway an outstanding task fmopv.

Reply:
So, how was this addressed? :)

Reply:
By ignoring it for now.

Reply:
It is sent as burst as we don't timestamp the bufferlist that we send, so there in no synchronization happenning after gccwbe.

What could be enhanced is synchronizatin upstream in the `clocksync` that is inside `webrtcbin` or the one before in `webrtcsink` itself.

Reply:
What do you mean "we do not timestamp the bufferlist that we send"? Bufferlists do not have a timestamp, sinks will synchronize (at the moment) by looking at the timestamp of the first buffer only, but there is no guarantee that this will remain true in the future.

Besides, webrtcbin could add elements that split up buffer lists downstream of the estimator for some reason (or might already do so, I haven't checked), so keeping the timestamps unchanged isn't very reassuring here :)

Reply:
Indeed, let me retimestamp buffers at this point.

Reply:
Right I got confused, let me remove pts from buffers individually

Also I did verify that we were not waiting on the clock and we weren't (because we sync on raw buffers already iirc)